[CI] FullRollingRestartIT testNoRebalanceOnRollingRestart failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/ccuezgxbievzg/tests/:server:internalClusterTest/org.elasticsearch.recovery.FullRollingRestartIT/testNoRebalanceOnRollingRestart

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.recovery.FullRollingRestartIT.testNoRebalanceOnRollingRestart" -Dtests.seed=41CBBA733A6A03C3 -Dtests.locale=ga-IE -Dtests.timezone=Asia/Dili -Druntime.java=17`

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.recovery.FullRollingRestartIT&tests.test=testNoRebalanceOnRollingRestart

**Failure excerpt:**
```
java.lang.AssertionError: relocated from: {node_t3}{EyyY-ftJSNudMYp5ZaNFdg}{og_ER-C-Sm-ycZCX9Ud44g}{node_t3}{127.0.0.1}{127.0.0.1:18854}{d} to: {node_t2}{omgXq9tVTbupq30bjCzm8w}{vlVodKB7QeKUdh92j9_WlA}{node_t2}{127.0.0.1}{127.0.0.1:18852}{d}-- 
before: 
cluster uuid: mEXdG06hSQGGtxtIaiKfZw [committed: true]
version: 13
state uuid: 47_pnsK9Q8SnGtede1V5HA
from_diff: false
meta data version: 8
   coordination_metadata:
      term: 1
      last_committed_config: VotingConfiguration{-dRfiVsgS0GqVuosHJ_Fhg}
      last_accepted_config: VotingConfiguration{-dRfiVsgS0GqVuosHJ_Fhg}
      voting tombstones: []
   [test/GasklFRaSBm1M9jduWKOJA]: v[6], mv[2], sv[1], av[1]
      0: p_term [1], isa_ids [PGuXQetKQgy4AbFMN4nHhA]
      1: p_term [1], isa_ids [RR60xWjgRriA_cV4Ne4ADw]
      2: p_term [1], isa_ids [NgxmsdRYSdWaNwua6QntbQ]
      3: p_term [1], isa_ids [Dr9qdp87SMGN4BKcFaQfpg]
      4: p_term [1], isa_ids [-zPxDnYZR0SEDiyIwu5XxQ]
      5: p_term [1], isa_ids [8-NhlnM4T8K6e3pyLc3dRw]
metadata customs:
   persistent_tasks: {"last_allocation_id":2,"tasks":[{"id":"health-node","task":{"health-node":{"params":{}}},"allocation_id":2,"assignment":{"executor_node":"omgXq9tVTbupq30bjCzm8w","explanation":""},"allocation_id_on_last_status_update":1}]}   index-graveyard: IndexGraveyard[[]]
nodes: 
   {node_t1}{pAVNxSNSSqiHdFDOgpad8Q}{S8nNtWoPRhyFow3vcFWdLA}{node_t1}{127.0.0.1}{127.0.0.1:18853}{d}
   {node_t2}{omgXq9tVTbupq30bjCzm8w}{FOQwYr4dT_mUODAHRXN9wg}{node_t2}{127.0.0.1}{127.0.0.1:18852}{d}
   {node_t3}{EyyY-ftJSNudMYp5ZaNFdg}{og_ER-C-Sm-ycZCX9Ud44g}{node_t3}{127.0.0.1}{127.0.0.1:18854}{d}
   {node_t0}{-dRfiVsgS0GqVuosHJ_Fhg}{1wrgT-RGSJKJtvuMmlx_1A}{node_t0}{127.0.0.1}{127.0.0.1:18851}{m}, local, master
routing_table (version 5):
-- index [[test/GasklFRaSBm1M9jduWKOJA]]
----shard_id [test][0]
--------[test][0], node[omgXq9tVTbupq30bjCzm8w], [P], s[STARTED], a[id=PGuXQetKQgy4AbFMN4nHhA], failed_attempts[0]
----shard_id [test][1]
--------[test][1], node[EyyY-ftJSNudMYp5ZaNFdg], [P], s[STARTED], a[id=RR60xWjgRriA_cV4Ne4ADw], failed_attempts[0]
----shard_id [test][2]
--------[test][2], node[pAVNxSNSSqiHdFDOgpad8Q], [P], s[STARTED], a[id=NgxmsdRYSdWaNwua6QntbQ], failed_attempts[0]
----shard_id [test][3]
--------[test][3], node[omgXq9tVTbupq30bjCzm8w], [P], s[STARTED], a[id=Dr9qdp87SMGN4BKcFaQfpg], failed_attempts[0]
----shard_id [test][4]
--------[test][4], node[EyyY-ftJSNudMYp5ZaNFdg], [P], s[STARTED], a[id=-zPxDnYZR0SEDiyIwu5XxQ], failed_attempts[0]
----shard_id [test][5]
--------[test][5], node[pAVNxSNSSqiHdFDOgpad8Q], [P], s[STARTED], a[id=8-NhlnM4T8K6e3pyLc3dRw], failed_attempts[0]

routing_nodes:
-----node_id[omgXq9tVTbupq30bjCzm8w][V]
--------[test][0], node[omgXq9tVTbupq30bjCzm8w], [P], s[STARTED], a[id=PGuXQetKQgy4AbFMN4nHhA], failed_attempts[0]
--------[test][3], node[omgXq9tVTbupq30bjCzm8w], [P], s[STARTED], a[id=Dr9qdp87SMGN4BKcFaQfpg], failed_attempts[0]
-----node_id[EyyY-ftJSNudMYp5ZaNFdg][V]
--------[test][1], node[EyyY-ftJSNudMYp5ZaNFdg], [P], s[STARTED], a[id=RR60xWjgRriA_cV4Ne4ADw], failed_attempts[0]
--------[test][4], node[EyyY-ftJSNudMYp5ZaNFdg], [P], s[STARTED], a[id=-zPxDnYZR0SEDiyIwu5XxQ], failed_attempts[0]
-----node_id[pAVNxSNSSqiHdFDOgpad8Q][V]
--------[test][2], node[pAVNxSNSSqiHdFDOgpad8Q], [P], s[STARTED], a[id=NgxmsdRYSdWaNwua6QntbQ], failed_attempts[0]
--------[test][5], node[pAVNxSNSSqiHdFDOgpad8Q], [P], s[STARTED], a[id=8-NhlnM4T8K6e3pyLc3dRw], failed_attempts[0]
---- unassigned
customs:
   health: org.elasticsearch.health.metadata.HealthMetadata@ff2c405a

  at __randomizedtesting.SeedInfo.seed([41CBBA733A6A03C3:497E64700C4572BD]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.assertTrue(Assert.java:41)
  at org.elasticsearch.recovery.FullRollingRestartIT.testNoRebalanceOnRollingRestart(FullRollingRestartIT.java:213)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/server/src/internalClusterTest/java/org/elasticsearch/recovery/FullRollingRestartIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/recovery/FullRollingRestartIT.java
@@ -199,7 +199,7 @@ public class FullRollingRestartIT extends ESIntegTestCase {
         ClusterState state = client().admin().cluster().prepareState().get().getState();
         RecoveryResponse recoveryResponse = client().admin().indices().prepareRecoveries("test").get();
         for (RecoveryState recoveryState : recoveryResponse.shardRecoveryStates().get("test")) {
-            assertTrue(
+            assertNotEquals(
                 "relocated "
                     + recoveryState.getShardId()
                     + " from: "
@@ -208,16 +208,17 @@ public class FullRollingRestartIT extends ESIntegTestCase {
                     + recoveryState.getTargetNode()
                     + "\n"
                     + state,
-                recoveryState.getRecoverySource().getType() != RecoverySource.Type.PEER || recoveryState.getPrimary() == false
+                recoveryState.getRecoverySource().getType(),
+                RecoverySource.Type.PEER
             );
         }
         internalCluster().restartRandomDataNode();
         ensureGreen();
-        client().admin().cluster().prepareState().get().getState();
+        client().admin().cluster().prepareState().get();
 
         recoveryResponse = client().admin().indices().prepareRecoveries("test").get();
         for (RecoveryState recoveryState : recoveryResponse.shardRecoveryStates().get("test")) {
-            assertTrue(
+            assertNotEquals(
                 "relocated "
                     + recoveryState.getShardId()
                     + " from: "
@@ -226,7 +227,8 @@ public class FullRollingRestartIT extends ESIntegTestCase {
                     + recoveryState.getTargetNode()
                     + "-- \nbefore: \n"
                     + state,
-                recoveryState.getRecoverySource().getType() != RecoverySource.Type.PEER || recoveryState.getPrimary() == false
+                recoveryState.getRecoverySource().getType(),
+                RecoverySource.Type.PEER
             );
         }
     }

--- a/server/src/internalClusterTest/java/org/elasticsearch/recovery/FullRollingRestartIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/recovery/FullRollingRestartIT.java
@@ -171,13 +171,12 @@ public class FullRollingRestartIT extends ESIntegTestCase {
         }
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/91441")
     public void testNoRebalanceOnRollingRestart() throws Exception {
         // see https://github.com/elastic/elasticsearch/issues/14387
         internalCluster().startMasterOnlyNode(Settings.EMPTY);
         internalCluster().startDataOnlyNodes(3);
         /**
-         * We start 3 nodes and a dedicated master. Restart on of the data-nodes and ensure that we got no relocations.
+         * We start 3 nodes and a dedicated master. Restart one of the data-nodes and ensure that we got no relocations.
          * Yet we have 6 shards 0 replica so that means if the restarting node comes back both other nodes are subject
          * to relocating to the restarting node since all had 2 shards and now one node has nothing allocated.
          * We have a fix for this to wait until we have allocated unallocated shards now so this shouldn't happen.
@@ -201,7 +200,14 @@ public class FullRollingRestartIT extends ESIntegTestCase {
         RecoveryResponse recoveryResponse = client().admin().indices().prepareRecoveries("test").get();
         for (RecoveryState recoveryState : recoveryResponse.shardRecoveryStates().get("test")) {
             assertTrue(
-                "relocated from: " + recoveryState.getSourceNode() + " to: " + recoveryState.getTargetNode() + "\n" + state,
+                "relocated "
+                    + recoveryState.getShardId()
+                    + " from: "
+                    + recoveryState.getSourceNode()
+                    + " to: "
+                    + recoveryState.getTargetNode()
+                    + "\n"
+                    + state,
                 recoveryState.getRecoverySource().getType() != RecoverySource.Type.PEER || recoveryState.getPrimary() == false
             );
         }
@@ -212,7 +218,14 @@ public class FullRollingRestartIT extends ESIntegTestCase {
         recoveryResponse = client().admin().indices().prepareRecoveries("test").get();
         for (RecoveryState recoveryState : recoveryResponse.shardRecoveryStates().get("test")) {
             assertTrue(
-                "relocated from: " + recoveryState.getSourceNode() + " to: " + recoveryState.getTargetNode() + "-- \nbefore: \n" + state,
+                "relocated "
+                    + recoveryState.getShardId()
+                    + " from: "
+                    + recoveryState.getSourceNode()
+                    + " to: "
+                    + recoveryState.getTargetNode()
+                    + "-- \nbefore: \n"
+                    + state,
                 recoveryState.getRecoverySource().getType() != RecoverySource.Type.PEER || recoveryState.getPrimary() == false
             );
         }

--- a/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java
@@ -32,8 +32,10 @@ import java.util.Set;
 import java.util.TreeMap;
 import java.util.TreeSet;
 import java.util.function.Predicate;
+import java.util.stream.Stream;
 
 import static java.util.stream.Collectors.toSet;
+import static java.util.stream.Collectors.toUnmodifiableSet;
 
 /**
  * Holds the desired balance and updates it as the cluster evolves.
@@ -62,7 +64,7 @@ public class DesiredBalanceComputer {
         final var routingAllocation = desiredBalanceInput.routingAllocation().mutableCloneForSimulation();
         final var routingNodes = routingAllocation.routingNodes();
         final var changes = routingAllocation.changes();
-        final var ignoredShards = desiredBalanceInput.ignoredShards();
+        final var ignoredShards = getIgnoredShardsWithDiscardedAllocationStatus(desiredBalanceInput.ignoredShards());
         final var knownNodeIds = routingAllocation.nodes().stream().map(DiscoveryNode::getId).collect(toSet());
         final var clusterInfoSimulator = new ClusterInfoSimulator(routingAllocation.clusterInfo());
 
@@ -90,7 +92,8 @@ public class DesiredBalanceComputer {
                 final var shardRouting = iterator.next();
                 if (shardRouting.primary() == primary) {
                     var lastAllocatedNodeId = shardRouting.unassignedInfo().getLastAllocatedNodeId();
-                    if (knownNodeIds.contains(lastAllocatedNodeId) || ignoredShards.contains(shardRouting) == false) {
+                    if (knownNodeIds.contains(lastAllocatedNodeId)
+                        || ignoredShards.contains(discardAllocationStatus(shardRouting)) == false) {
                         shardRoutings.computeIfAbsent(shardRouting.shardId(), ShardRoutings::new).unassigned().add(shardRouting);
                     } else {
                         iterator.removeAndIgnore(UnassignedInfo.AllocationStatus.NO_ATTEMPT, changes);
@@ -216,10 +219,11 @@ public class DesiredBalanceComputer {
                 // Not the first iteration, so every remaining unassigned shard has been ignored, perhaps due to throttling. We must bring
                 // them all back out of the ignored list to give the allocator another go...
                 routingNodes.unassigned().resetIgnored();
+
                 // ... but not if they're ignored because they're out of scope for allocation
                 for (final var iterator = routingNodes.unassigned().iterator(); iterator.hasNext();) {
                     final var shardRouting = iterator.next();
-                    if (ignoredShards.contains(shardRouting)) {
+                    if (ignoredShards.contains(discardAllocationStatus(shardRouting))) {
                         iterator.removeAndIgnore(UnassignedInfo.AllocationStatus.NO_ATTEMPT, changes);
                     }
                 }
@@ -314,5 +318,36 @@ public class DesiredBalanceComputer {
         private ShardRoutings(ShardId ignored) {
             this(new ArrayList<>(), new ArrayList<>());
         }
+    }
+
+    private static Set<ShardRouting> getIgnoredShardsWithDiscardedAllocationStatus(List<ShardRouting> ignoredShards) {
+        return ignoredShards.stream()
+            .flatMap(shardRouting -> Stream.of(shardRouting, discardAllocationStatus(shardRouting)))
+            .collect(toUnmodifiableSet());
+    }
+
+    /**
+     * AllocationStatus is discarded as it might come from GatewayAllocator and not be present in corresponding routing table
+     */
+    private static ShardRouting discardAllocationStatus(ShardRouting shardRouting) {
+        return shardRouting.updateUnassigned(discardAllocationStatus(shardRouting.unassignedInfo()), shardRouting.recoverySource());
+    }
+
+    private static UnassignedInfo discardAllocationStatus(UnassignedInfo info) {
+        if (info.getLastAllocationStatus() == UnassignedInfo.AllocationStatus.NO_ATTEMPT) {
+            return info;
+        }
+        return new UnassignedInfo(
+            info.getReason(),
+            info.getMessage(),
+            info.getFailure(),
+            info.getNumFailedAllocations(),
+            info.getUnassignedTimeInNanos(),
+            info.getUnassignedTimeInMillis(),
+            info.isDelayed(),
+            UnassignedInfo.AllocationStatus.NO_ATTEMPT,
+            info.getFailedNodeIds(),
+            info.getLastAllocatedNodeId()
+        );
     }
 }

--- a/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java
@@ -32,7 +32,6 @@ import java.util.Set;
 import java.util.TreeMap;
 import java.util.TreeSet;
 import java.util.function.Predicate;
-import java.util.stream.Stream;
 
 import static java.util.stream.Collectors.toSet;
 import static java.util.stream.Collectors.toUnmodifiableSet;
@@ -321,9 +320,7 @@ public class DesiredBalanceComputer {
     }
 
     private static Set<ShardRouting> getIgnoredShardsWithDiscardedAllocationStatus(List<ShardRouting> ignoredShards) {
-        return ignoredShards.stream()
-            .flatMap(shardRouting -> Stream.of(shardRouting, discardAllocationStatus(shardRouting)))
-            .collect(toUnmodifiableSet());
+        return ignoredShards.stream().map(DesiredBalanceComputer::discardAllocationStatus).collect(toUnmodifiableSet());
     }
 
     /**

--- a/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceInput.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceInput.java
@@ -9,13 +9,9 @@
 package org.elasticsearch.cluster.routing.allocation.allocator;
 
 import org.elasticsearch.cluster.routing.ShardRouting;
-import org.elasticsearch.cluster.routing.UnassignedInfo;
 import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
 
-import java.util.Set;
-import java.util.stream.Stream;
-
-import static java.util.stream.Collectors.toUnmodifiableSet;
+import java.util.List;
 
 /**
  * The input to the desired balance computation.
@@ -26,45 +22,13 @@ import static java.util.stream.Collectors.toUnmodifiableSet;
  * @param routingAllocation a copy of (the immutable parts of) the context for the allocation decision process
  * @param ignoredShards     a list of the shards for which earlier allocators have claimed responsibility
  */
-public record DesiredBalanceInput(long index, RoutingAllocation routingAllocation, Set<ShardRouting> ignoredShards) {
+public record DesiredBalanceInput(long index, RoutingAllocation routingAllocation, List<ShardRouting> ignoredShards) {
 
     public static DesiredBalanceInput create(long index, RoutingAllocation routingAllocation) {
         return new DesiredBalanceInput(
             index,
             routingAllocation.immutableClone(),
-            getIgnoredShardsWithDiscardedAllocationStatus(routingAllocation)
-        );
-    }
-
-    private static Set<ShardRouting> getIgnoredShardsWithDiscardedAllocationStatus(RoutingAllocation routingAllocation) {
-        return routingAllocation.routingNodes()
-            .unassigned()
-            .ignored()
-            .stream()
-            .flatMap(
-                shardRouting -> Stream.of(
-                    shardRouting,
-                    shardRouting.updateUnassigned(discardAllocationStatus(shardRouting.unassignedInfo()), shardRouting.recoverySource())
-                )
-            )
-            .collect(toUnmodifiableSet());
-    }
-
-    /**
-     * AllocationStatus is discarded as it might come from GatewayAllocator and not be present in corresponding routing table
-     */
-    private static UnassignedInfo discardAllocationStatus(UnassignedInfo info) {
-        return new UnassignedInfo(
-            info.getReason(),
-            info.getMessage(),
-            info.getFailure(),
-            info.getNumFailedAllocations(),
-            info.getUnassignedTimeInNanos(),
-            info.getUnassignedTimeInMillis(),
-            info.isDelayed(),
-            UnassignedInfo.AllocationStatus.NO_ATTEMPT,
-            info.getFailedNodeIds(),
-            info.getLastAllocatedNodeId()
+            List.copyOf(routingAllocation.routingNodes().unassigned().ignored())
         );
     }
 }

--- a/server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java
@@ -21,6 +21,7 @@ import org.elasticsearch.cluster.node.DiscoveryNodeRole;
 import org.elasticsearch.cluster.node.DiscoveryNodes;
 import org.elasticsearch.cluster.routing.AllocationId;
 import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.cluster.routing.RecoverySource;
 import org.elasticsearch.cluster.routing.RoutingChangesObserver;
 import org.elasticsearch.cluster.routing.RoutingTable;
@@ -120,9 +121,9 @@ public class DesiredBalanceComputerTests extends ESTestCase {
 
     public void testIgnoresOutOfScopePrimaries() {
         var desiredBalanceComputer = createDesiredBalanceComputer();
-        var clusterState = createInitialClusterState(3);
+        var clusterState = mutateAllocationStatuses(createInitialClusterState(3));
         var index = clusterState.metadata().index(TEST_INDEX).getIndex();
-        var primaryShard = clusterState.routingTable().index(TEST_INDEX).shard(0).primaryShard();
+        var primaryShard = mutateAllocationStatus(clusterState.routingTable().index(TEST_INDEX).shard(0).primaryShard());
 
         var desiredBalance = desiredBalanceComputer.compute(
             DesiredBalance.INITIAL,
@@ -135,7 +136,18 @@ public class DesiredBalanceComputerTests extends ESTestCase {
             desiredBalance,
             Map.of(
                 new ShardId(index, 0),
-                new ShardAssignment(Set.of(), 2, 2, 2),
+                new ShardAssignment(
+                    Set.of(),
+                    2,
+                    2,
+                    clusterState.routingTable()
+                        .index(TEST_INDEX)
+                        .shard(0)
+                        .replicaShards()
+                        .get(0)
+                        .unassignedInfo()
+                        .getLastAllocationStatus() == UnassignedInfo.AllocationStatus.DECIDERS_NO ? 1 : 2
+                ),
                 new ShardId(index, 1),
                 new ShardAssignment(Set.of("node-0", "node-1"), 2, 0, 0)
             )
@@ -144,10 +156,11 @@ public class DesiredBalanceComputerTests extends ESTestCase {
 
     public void testIgnoresOutOfScopeReplicas() {
         var desiredBalanceComputer = createDesiredBalanceComputer();
-        var clusterState = createInitialClusterState(3);
+        var clusterState = mutateAllocationStatuses(createInitialClusterState(3));
 
         var index = clusterState.metadata().index(TEST_INDEX).getIndex();
-        var replicaShard = clusterState.routingTable().index(TEST_INDEX).shard(0).replicaShards().get(0);
+        var originalReplicaShard = clusterState.routingTable().index(TEST_INDEX).shard(0).replicaShards().get(0);
+        var replicaShard = mutateAllocationStatus(originalReplicaShard);
 
         var desiredBalance = desiredBalanceComputer.compute(
             DesiredBalance.INITIAL,
@@ -160,7 +173,12 @@ public class DesiredBalanceComputerTests extends ESTestCase {
             desiredBalance,
             Map.of(
                 new ShardId(index, 0),
-                new ShardAssignment(Set.of("node-0"), 2, 1, 1),
+                new ShardAssignment(
+                    Set.of("node-0"),
+                    2,
+                    1,
+                    originalReplicaShard.unassignedInfo().getLastAllocationStatus() == UnassignedInfo.AllocationStatus.DECIDERS_NO ? 0 : 1
+                ),
                 new ShardId(index, 1),
                 new ShardAssignment(Set.of("node-0", "node-1"), 2, 0, 0)
             )
@@ -579,7 +597,7 @@ public class DesiredBalanceComputerTests extends ESTestCase {
         var input = new DesiredBalanceInput(
             randomInt(),
             routingAllocationWithDecidersOf(clusterState, ClusterInfo.EMPTY, Settings.EMPTY),
-            Set.of()
+            List.of()
         );
         var desiredBalance = new DesiredBalanceComputer(new BalancedShardsAllocator(Settings.EMPTY)).compute(
             DesiredBalance.INITIAL,
@@ -749,7 +767,7 @@ public class DesiredBalanceComputerTests extends ESTestCase {
 
         var desiredBalance = new DesiredBalanceComputer(new BalancedShardsAllocator(settings)).compute(
             initial,
-            new DesiredBalanceInput(randomInt(), routingAllocationWithDecidersOf(clusterState, clusterInfo, settings), Set.of()),
+            new DesiredBalanceInput(randomInt(), routingAllocationWithDecidersOf(clusterState, clusterInfo, settings), List.of()),
             queue(),
             input -> true
         );
@@ -793,6 +811,50 @@ public class DesiredBalanceComputerTests extends ESTestCase {
             .metadata(Metadata.builder().put(indexMetadata, true))
             .routingTable(RoutingTable.builder().addAsNew(indexMetadata))
             .build();
+    }
+
+    static ClusterState mutateAllocationStatuses(ClusterState clusterState) {
+        final var routingTableBuilder = RoutingTable.builder();
+        for (final var indexRoutingTable : clusterState.routingTable()) {
+            final var indexRoutingTableBuilder = new IndexRoutingTable.Builder(indexRoutingTable.getIndex());
+            for (int shardId = 0; shardId < indexRoutingTable.size(); shardId++) {
+                final var shardRoutingTable = indexRoutingTable.shard(shardId);
+                final var shardRoutingTableBuilder = new IndexShardRoutingTable.Builder(shardRoutingTable.shardId());
+                for (int shardCopy = 0; shardCopy < shardRoutingTable.size(); shardCopy++) {
+                    shardRoutingTableBuilder.addShard(mutateAllocationStatus(shardRoutingTable.shard(shardCopy)));
+                }
+                indexRoutingTableBuilder.addIndexShard(shardRoutingTableBuilder);
+            }
+            routingTableBuilder.add(indexRoutingTableBuilder);
+        }
+        return ClusterState.builder(clusterState).routingTable(routingTableBuilder).build();
+    }
+
+    private static ShardRouting mutateAllocationStatus(ShardRouting shardRouting) {
+        if (shardRouting.unassigned()) {
+            var unassignedInfo = shardRouting.unassignedInfo();
+            return shardRouting.updateUnassigned(
+                new UnassignedInfo(
+                    unassignedInfo.getReason(),
+                    unassignedInfo.getMessage(),
+                    unassignedInfo.getFailure(),
+                    unassignedInfo.getNumFailedAllocations(),
+                    unassignedInfo.getUnassignedTimeInNanos(),
+                    unassignedInfo.getUnassignedTimeInMillis(),
+                    unassignedInfo.isDelayed(),
+                    randomFrom(
+                        UnassignedInfo.AllocationStatus.DECIDERS_NO,
+                        UnassignedInfo.AllocationStatus.NO_ATTEMPT,
+                        UnassignedInfo.AllocationStatus.DECIDERS_THROTTLED
+                    ),
+                    unassignedInfo.getFailedNodeIds(),
+                    unassignedInfo.getLastAllocatedNodeId()
+                ),
+                shardRouting.recoverySource()
+            );
+        } else {
+            return shardRouting;
+        }
     }
 
     private static DiscoveryNode createDiscoveryNode(String id, Set<DiscoveryNodeRole> roles) {
@@ -843,7 +905,7 @@ public class DesiredBalanceComputerTests extends ESTestCase {
     }
 
     private static DesiredBalanceInput createInput(ClusterState clusterState, ShardRouting... ignored) {
-        return new DesiredBalanceInput(randomInt(), routingAllocationOf(clusterState), Set.of(ignored));
+        return new DesiredBalanceInput(randomInt(), routingAllocationOf(clusterState), List.of(ignored));
     }
 
     private static RoutingAllocation routingAllocationOf(ClusterState clusterState) {

